### DISCUSSION TITLE
Remove Intel brew on macOS Monterey (macOS 12)

### DISCUSSION THREAD

**Discussion by louy2** [Homebrew/discussions#2408] [Everyday usage] [answered]
On Macbook Pro 2020 with Apple M1, I have a copy of Intel homebrew migrated from my old Macbook, so I followed the official instruction at the time to install a separate copy of M1 homebrew, and alias the Intel copy to `ibrew`.

I updated to macOS Monterey last week, and today I've found out that `ibrew` fails with `MacOSVersionError`:

```
/usr/local/Homebrew/Library/Homebrew/os/mac/version.rb:28:in `block in from_symbol': unknown or unsupported macOS version: :dunno (MacOSVersionError)
	from /usr/local/Homebrew/Library/Homebrew/os/mac/version.rb:28:in `fetch'
	from /usr/local/Homebrew/Library/Homebrew/os/mac/version.rb:28:in `from_symbol'
	from /usr/local/Homebrew/Library/Homebrew/os/mac.rb:28:in `version'
	from /usr/local/Homebrew/Library/Homebrew/os/mac.rb:65:in `prerelease?'
	from /usr/local/Homebrew/Library/Homebrew/os.rb:38:in `<module:OS>'
	from /usr/local/Homebrew/Library/Homebrew/os.rb:7:in `<top (required)>'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'
	from /usr/local/Homebrew/Library/Homebrew/global.rb:47:in `<top (required)>'
	from /usr/local/Homebrew/Library/Homebrew/brew.rb:31:in `require_relative'
	from /usr/local/Homebrew/Library/Homebrew/brew.rb:31:in `<main>'
```

The reason is simple enough: my Intel copy of homebrew is never updated, and it cannot handle the new version number. I am not sure if I can still update it. I have moved to use the M1 brew mostly, and my Intel brew has caused some linking errors, so I want to remove it completely, but without running it I don't know how to remove it.

How may I remove my Intel homebrew now?

**Comment by carlocab** [comment, marked as answer]
You can use the uninstall script at https://github.com/Homebrew/install#uninstall-homebrew. You don't want to run the snippet shown there (that will likely uninstall your `/opt/homebrew` installation). Instead, you want to download the script, make it executable, then run it with the flag `--path=/usr/local`. You may need to run it with `sudo`.

**Reply by dpkonofa** [reply]
Just a note of caution. If you run this *without* sudo and it fails, you won't be able to run it with sudo again afterwards and will be left with orphaned files. I ran it without sudo and it threw the error:

`rmdir: /usr/local: Operation not permitted
Warning: Failed during: /usr/bin/sudo rmdir /usr/local
Warning: Homebrew partially uninstalled (but there were steps that failed)!
To finish uninstalling rerun this script with ``sudo``.
The following possible Homebrew files were not deleted:
/usr/local/.com.apple.installer.keep
/usr/local/Frameworks/
/usr/local/Homebrew/
/usr/local/anaconda3/
/usr/local/bin/
/usr/local/etc/
/usr/local/include/
/usr/local/lib/
/usr/local/libexec/
/usr/local/opt/
/usr/local/remotedesktop/
/usr/local/sbin/
/usr/local/share/
/usr/local/var/
You may wish to remove them yourself.`

Not sure if that means to delete the entire /usr/local directory or not (since it looks like that's what it was attempting to do) or re-install and try again.

**Reply by agrakhov** [reply]
Hey, @dpkonofa, I'm having the same issue: run the uninstall command, which deleted bin/brew but failed in the middle, exactly like in your case.
Did you end up removing the `/usr/local` folder completely?

**Reply by dpkonofa** [reply]
@agrakhov - I did. I ran the whole install process on another machine in a clean install directory and, as far as I could tell, /usr/local wasn't used for anything else. I deleted everything and nothing seems to have broken. Reinstalling brew then installed the M1 version.

**Reply by DSep** [reply]
A note for people following this answer in the future:

As someone who has other things like texlive (MacTex) located in `/usr/local`, I would recommend at minimum looking inside the folder and ensuring you don't have anything else you need before running the uninstall script or deleting the whole thing (which are essentially the same action).

I ended up manually deleting things that were HomeBrew related. I hope they can make a more careful uninstall script (or maybe I'll come back and fix it when I have some time).

**Reply by henno** [reply]
```
henno@Hennos-MBP-3 ~ % wget https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh
zsh: bad CPU type in executable: wget
henno@Hennos-MBP-3 ~ % curl https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh -o uninstall.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 11306  100 11306    0     0  41133      0 --:--:-- --:--:-- --:--:-- 41874
henno@Hennos-MBP-3 ~ % chmod +x uninstall.sh 
henno@Hennos-MBP-3 ~ % sudo ./uninstall.sh --path=/usr/local
Password:
Warning: This script will remove:
/usr/local/Caskroom/
/usr/local/Cellar/
/usr/local/bin/brew -> /usr/local/bin/brew
Are you sure you want to uninstall Homebrew? This will remove your installed packages! [y/N] y
==> Removing Homebrew installation...
==> Removing empty directories...
==> /usr/bin/sudo /usr/bin/find /usr/local/bin /usr/local/etc /usr/local/include /usr/local/lib /usr/local/opt /usr/local/sbin /usr/local/share /usr/local/var /usr/local/Homebrew /usr/local/Frameworks -name .DS_Store -delete
==> /usr/bin/sudo /usr/bin/find /usr/local/bin /usr/local/etc /usr/local/include /usr/local/lib /usr/local/opt /usr/local/sbin /usr/local/share /usr/local/var /usr/local/Homebrew /usr/local/Frameworks -depth -type d -empty -exec rmdir {} ;
==> /usr/bin/sudo rmdir /usr/local
rmdir: /usr/local: Operation not permitted
Warning: Failed during: /usr/bin/sudo rmdir /usr/local
Warning: Homebrew partially uninstalled (but there were steps that failed)!
To finish uninstalling rerun this script with `sudo`.
The following possible Homebrew files were not deleted:
/usr/local/Frameworks/
/usr/local/Homebrew/
/usr/local/bin/
/usr/local/etc/
/usr/local/include/
/usr/local/lib/
/usr/local/man/
/usr/local/opt/
/usr/local/sbin/
/usr/local/share/
/usr/local/var/
You may wish to remove them yourself.

henno@Hennos-MBP-3 ~ % 
```
I'm not sure if this is was what was supposed to happen and if I should remove these directories myself or not.

**Reply by osalbahr** [reply]
@henno the output is expected because these directories require `sudo` to remove them. You can look into these directories and decide if you want to remove them manually. And, imho, it is safe to just keep them as-is. The uninstall script should have already removed Homebrew, just not the directories above as some might be used by other programs.

**Reply by henno** [reply]
I did use sudo.

**Reply by osalbahr** [reply]
Sorry, I just realized that you already did run the script as sudo but the behavior contradicts with the message that you got. I recommend opening an issue at https://github.com/Homebrew/install/issues

**Reply by henno** [reply]
In order to reproduce the bug, one must migrate from an Intel Mac which is doable but would require a spare Mac and a Time Machine backup and lots of time. I just installed brew again by executing `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"` and everything seems to be normal now.

**Reply by osalbahr** [reply]
@henno I’m curious, now that everything is normal, what happens if you re-run the uninstall script as sudo? I expect the same to happen.

Also, I think `cp -r` would be enough to simulate what happened with Time Machine. Someone can, for example, zip these directories and transfer them. Or, just manually creating the directories (with sudo) to simulate as if it were previously installed in an Intel mac. Regardless, I think this is a bug because you did use sudo and --path. And it is up to whoever sees the issue whether it’s worth the effort for them. That repo has many watchers and people periodically checking the issues.

I recommend opening an issue because I think it is a bug and the answer was selected by the OP which means it probably fit their need.

- - -

Side note: check `brew help bundle` for an easy way to get a list of all that you installed and be able to easily reinstall them. I think the original issue (wget) might be fixable by just `brew reinstall wget`. It might be already too late for that now unless you go back to the backup, just for future reference.

**Reply by yaohwu** [reply]
--path=/usr/local
help me a lot

**Comment by Bo98** [comment]
> I am not sure if I can still update it.

`ibrew update` should work. It does all updating before it even starts the Ruby part.

(But updating is not necessary if you are planning to uninstall.)

> so I want to remove it completely, but without running it I don't know how to remove it.

```
arch -x86_64 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/uninstall.sh)"
```

It will show what paths it is removing (it should be the ones under `/usr/local`) and ask for a confirmation.

**Reply by louy2** [reply]
Thank you and `ibrew update` worked, but the paths the command removes is under `/opt/homebrew` so I have canceled it.

Between the question and now I remembered I need the Intel one to manage my x86_64 Java installations so I need to keep the Intel brew for now.

**Reply by louy2** [reply]
I realized I can't use the Intel one to manage it anyway. XD

**Reply by victor-kwan-zh** [reply]
It works, thx...
